Implement mbedtls_pk_copy_from_psa

valeriosetti opened the pull request:
## Description

Adding `mbedtls_pk_copy_from_psa` as described in https://github.com/Mbed-TLS/mbedtls/blob/development/docs/architecture/psa-migration/psa-legacy-bridges.md#api-to-copy-a-psa-key-to-a-pk-context

Depends on ~https://github.com/Mbed-TLS/mbedtls/pull/8740~, ~#8824~

Resolves #8709

## PR checklist

- [x] **changelog** provided
- [x] **backport** not required: new feature
- [x] **tests** provided

valeriosetti commented:
Opening this PR early to get some feedback from those who worked on `psa-legacy-bridges` document (@mpg & @gilles-peskine-arm).
Assuming my implementation of `mbedtls_pk_copy_from_psa` is somehow close to the requirements, there are some functions from the `pkparse` module that should be made available internally to the library in order to ease the filling of the PK context with EC/RSA data (see https://github.com/Mbed-TLS/mbedtls/pull/8774/commits/e0607dfb04919cd9aaa7b61e9154980dfdeae77a). As for RSA these should be already moved to `rsa.c` module in https://github.com/Mbed-TLS/mbedtls/pull/8740. What should we do for EC ones?
I think that keeping the declaration in `pk_internal.c` and having the implementation in `pk.c` should be fine, wdyt?

mpg commented on include/mbedtls/pk.h:
```diff
@@ -28,7 +28,7 @@
 #endif
 
-#if defined(MBEDTLS_USE_PSA_CRYPTO)
+#if defined(MBEDTLS_USE_PSA_CRYPTO) || defined(MBEDTLS_PSA_CRYPTO_CLIENT)
```
That's redundant, we could remove `USE_PSA_CRYPTO` then.

mpg commented on library/pk.c:
```diff
@@ -29,7 +29,7 @@
 #endif
 
-#if defined(MBEDTLS_USE_PSA_CRYPTO)
+#if defined(MBEDTLS_USE_PSA_CRYPTO) || defined(MBEDTLS_PSA_CRYPTO_CLIENT)
```
Same.

gilles-peskine-arm commented:
> there are some functions from the pkparse module that should be made available internally to the library in order to ease the filling of the PK context with EC/RSA data (see e0607df). As for RSA these should be already moved to rsa.c module in #8740. What should we do for EC ones?

I would have expected that for RSA, but it's (about to be) resolved by #8740 so no pkparse function should be needed there.

For EC, the only parsing is for public keys, and I would have expected `mbedtls_ecp_point_read_binary` to be sufficient, no need for any pkparse code. Where's the gap?

mpg commented on library/pk.c:
```diff
@@ -877,6 +877,102 @@ mbedtls_pk_type_t mbedtls_pk_get_type(const mbedtls_pk_context *ctx)
+    psa_algorithm_t alg_type;
+    size_t key_bits;
+    /* Use a buffer size large enough to contain either a key pair or public key. */
+    unsigned char exp_key[PSA_EXPORT_KEY_PAIR_MAX_SIZE];
```
Are we sure this is always large enough? For ECC keys, a key pair (unlike the name suggests) is exported as only the private key (one integer), so it's actually smaller than just a public key for Short Weierstrass curves (two integers + 1 byte). I seem to remember having this discussion before so we might have a macro for that already.

@gilles-peskine-arm This is arguably a bit of a gotcha in the PSA Crypto API, as it seems natural to assume a "pair" is larger than one of its components. Should we actually add a macro like `PSA_EXPORT_KEY_PAIR_OR_PUBLIC_MAX_SIZE` (or a better name) to the official PSA API?

gilles-peskine-arm replied:
> This is arguably a bit of a gotcha in the PSA Crypto API, as it seems natural to assume a "pair" is larger than one of its components.

Indeed. The key pair object is “larger” in terms of carrying strictly more information, but the representation doesn't have to be larger. Perhaps “key pair” objects should have been called “private key” objects, but that also as downsides, in particular that it's not obvious that every private key object implicitly includes a public key.

> `PSA_EXPORT_KEY_PAIR_OR_PUBLIC_MAX_SIZE`

I suppose we could. It's easy enough to implement. I don't see a lot of value to it: it's uncommon that you'd know you have a key pair or a public key, but you don't know which. I guess that would only come up in key management APIs, which is sort of what we're doing here.

gilles-peskine-arm replied:
I'm working on `mbedtls_pk_import_into_psa` and naturally I want to do exactly the same thing. What we're doing is a bit niche, but for that niche, `PSA_EXPORT_KEY_PAIR_OR_PUBLIC_MAX_SIZE` would indeed be convenient.

mpg reviewed (commented):
(Partial review, stopped in the middle of 95da550b700e582efe0fdf52caf2d22460d04310. Just releasing pending comments before I leave for today.)

mpg commented:
> For EC, the only parsing is for public keys, and I would have expected `mbedtls_ecp_point_read_binary` to be sufficient, no need for any pkparse code. Where's the gap?

I think the gap is support for `USE_PSA_EC_DATA`. The functions Valerio wants to use are an abstraction layer for that.

> I think that keeping the declaration in pk_internal.c and having the implementation in pk.c should be fine, wdyt?

Makes sense to me. I think we may want to make this PR depend on #8740 for the RSA functions, though.

valeriosetti commented:
> > For EC, the only parsing is for public keys, and I would have expected `mbedtls_ecp_point_read_binary` to be sufficient, no need for any pkparse code. Where's the gap?
> 
> I think the gap is support for `USE_PSA_EC_DATA`. The functions Valerio wants to use are an abstraction layer for that.

Exactly: `pk_ecc_set_key`, `pk_ecc_set_pubkey` and `pk_ecc_set_pubkey_from_prv` nicely handle `MBEDTLS_PK_USE_PSA_EC_DATA` vs ECP way of storing data.

valeriosetti commented:
@gilles-peskine-arm @mpg 
I'm still trying to make the CI fully happy, but I suspect I might still have some issue with `test_psa_crypto_client` in the end. The problem is that I guarded the newly added `mbedtls_pk_copy_from_psa()` with `CRYPTO_CLIENT`, but as we know this is something that is not fully supported yet. As a consequence I think that I have to revert that `CRYPTO_CLIENT` guard with `CRYPTO_C` as it was recently done in https://github.com/Mbed-TLS/mbedtls/pull/8738.

gilles-peskine-arm commented:
I don't have time to dig into the details right now, but Manuel and I were chatting about this the other day and we were thinking that we might need to split `test_psa_crypto_client` into what can be linked and what can't. https://github.com/Mbed-TLS/mbedtls/issues/8710#issuecomment-1933825279 If you have time, could you explore this and see if it makes sense?

mpg commented:
> As a consequence I think that I have to revert that `CRYPTO_CLIENT` guard with `CRYPTO_C` as it was recently done in #8738.

Yes, that's the strategy I was going to suggest: do one thing at a time. First, implement the function (here) and then work on guards to make it available with `CRPYTO_CLIENT && !CRYPTO_C` builds (in a follow-up PR).

mpg commented on tests/suites/test_suite_pk.function:
```diff
@@ -1924,3 +1968,202 @@ exit:
+    pub_key_id = pk_psa_pub_key_from_priv(priv_key_id, pub_key_type, key_usage, key_alg, key_bits);
+    TEST_EQUAL(mbedtls_pk_copy_from_psa(pub_key_id, &pk_ctx2), 0);
+
+    /* Test sign/verify with the following parttern:
```
Typo: paRttern

mpg commented on tests/suites/test_suite_pk.function:
```diff
@@ -1924,3 +1968,202 @@ exit:
+     * - Verify from the same PK context used for signature.
+     * - Verify with the PK context generated using public key.
+     */
+    if ((PSA_ALG_IS_RSA_OAEP(key_alg) || PSA_ALG_IS_RSA_PSS(key_alg))) {
```
Note: ideally we want tests with the non-ext functions in this case as well. I think we still need a bit more discussion about what should be possible with what original algs (and what it should do).

mpg commented on library/pk_ecc.c:
```diff
@@ -0,0 +1,304 @@
+ *      ECC setters
+ *
+ * 1. This is an abstraction layer around MBEDTLS_PK_USE_PSA_EC_DATA:
+ *    this macro will not appear outside this section.
```
Note: this line no longer makes sense now that this is out of `pkparse.c`.

mpg commented on library/pk_ecc.c:
```diff
@@ -0,0 +1,304 @@
+#include "common.h"
```
We need the usual copyright+licence lines. Probably should move the description here too ("ECC setters" below).

mpg commented on tests/scripts/all.sh:
```diff
@@ -1531,6 +1531,28 @@ component_test_sw_inet_pton () {
+    scripts/config.py unset MBEDTLS_PK_PARSE_C
+    scripts/config.py unset MBEDTLS_PK_WRITE_C
+
+    # Disable features that re-enable PK_PARSE_C
```
This should no longer be necessary thanks to #8740.

mpg reviewed (commented):
(Just a first pass, getting myself familiar with this.)

mpg commented on library/pk.c:
```diff
@@ -1080,6 +1082,132 @@ mbedtls_pk_type_t mbedtls_pk_get_type(const mbedtls_pk_context *ctx)
+            }
+            ret = mbedtls_pk_ecc_set_pubkey_from_prv(pk, exp_key, exp_key_len,
+                                                     mbedtls_psa_get_random,
+                                                     MBEDTLS_PSA_RANDOM_STATE);
```
To be clear: nothing to do here. This was just a note for the future.

mpg commented on tests/suites/test_suite_pk.function:
```diff
@@ -1924,3 +1968,202 @@ exit:
 }
 /* END_CASE */
+
+/* BEGIN_CASE depends_on:MBEDTLS_PSA_CRYPTO_CLIENT*/
```
Again, nothing to do here, this was just FYI. In a future PR (pre-3.6) we want to change back guards from `_C` to `_CLIENT` in the library, but we'll want to keep `_C` in tests for the time being (until after 3.6).

mpg commented:
@gilles-peskine-arm As I said in a comment above, I think we need to discuss a bit more what functions it should be possible to call successfully and what they'll do depending on the original alg. (This is just for RSA keys, I think for ECC it's all clear.)

Here's a suggestion to get the discussion started.

1. **Original alg: PKCS#1 v1.5 (crypt or sign)**
   a. `decrypt`, `encrypt`: works and uses v1.5 encryption
   b. `sign`, `verify`: works and uses v1.5 signatures
   c. `sign_ext`, `verify_ext`: works and uses whatever algorithm was passed (v1.5 or PSS)
2. **Original alg: v2.1 (OAEP or PSS) with a concrete hash**
   a. `decrypt`, `encrypt`: errors out: PK does not support OAEP
   b. `sign`, `verify`: ??? (works and use PSS signatures with the original PSA hash?)
   c. `sign_ext`, `verify_ext`: ?? works and uses whatever algorithm (and hash) was passed (v1.5 or PSS) - but what if `NONE` was passed as `md_alg` for `sign_ext`?
3. **Original alg: v2.1 (OAEP or PSS) with a `ANY_HASH`**
   a. `decrypt`, `encrypt`: errors out: PK does not support OAEP
   b. `sign`, `verify`: ??? (errors out: needs a hash to be specified? or depends on whether a hash was passed?)
   c. `sign_ext`, `verify_ext`: ?? works and uses whatever algorithm (and hash) was passed (v1.5 or PSS) - but what if `NONE` was passed as `md_alg` for `sign_ext`?

**Doubts about 2b and 3b**

Reading again the documentation of `pk_sign`, the following seem of interest:
```
 * \note            For RSA keys, the default padding type is PKCS#1 v1.5.
 *                  There is no interface in the PK module to make RSASSA-PSS
 *                  signatures yet.
```
The 2nd sentence seems outdated (we have `pk_sign_ext()` now) but what about the first? 

The documentation of `pk_verify()` is similarly unhelpful:
```
 * \note            For RSA keys, the default padding type is PKCS#1 v1.5.
 *                  Use \c mbedtls_pk_verify_ext( MBEDTLS_PK_RSASSA_PSS, ... )
 *                  to verify RSASSA_PSS signatures.
```

What are these functions supposed to do with a context where `mbedtls_rsa_set_padding(mbedtls_pk_rsa(...), MBEDTLS_RSA_PKCS_V21, ...)` was called? That's unclear to me from the documentation. Even the intent is unclear to me from memories. Also, I doubt we are testing any of this.

**Doubts about 3c (and 2c?)**

The other problem I see is that both `sign()` and even `sign_ext()` don't require the use to specify a hash, as they both have the following in their documentation:
```
 * \note            For RSA, md_alg may be MBEDTLS_MD_NONE if hash_len != 0.
```
Similarly, `pk_verify()` allows the hash to be left unspecified:
```
 *                  This can be #MBEDTLS_MD_NONE if the signature algorithm
 *                  does not rely on a hash algorithm (non-deterministic
 *                  ECDSA, RSA PKCS#1 v1.5).
```
And so does `pk_verify_ext()`:
```
 * \note            md_alg may be MBEDTLS_MD_NONE, only if hash_len != 0
```
but here maybe we're saved by the `mbedtls_pk_rsassa_pss_options` which specifies the MGF1 hash?

**General thoughts**

I wonder if we're trying to do too much here by carrying over as much information as possible from PSA, into a PK module that was not really designed for that.

@gilles-peskine-arm wdyt?

gilles-peskine-arm commented:
> I wonder if we're trying to do too much here by carrying over as much information as possible from PSA, into a PK module that was not really designed for that.

Maybe. We could make `copy_from_psa` equivalent to exporting and parsing the key material, ignoring the key policy completely. This would be `psa_export_key` followed by `mbedtls_pk_parse_key` or `mbedtls_pk_parse_pubkey`, except without requiring extra work to add metadata indicating the curve for EC keys.

----

> 1\. PKCS#1 v1.5 (crypt or sign)

V1.5 crypt doesn't use a hash. v1.5 signature optionally uses a hash. The pk and rsa modules ignore the `hash_id` field in the RSA context when doing a v1.5 signature. So we don't have a choice here anyway. (Except the obvious choice of `rsa->padding` which we obviously leave at the default value `MBEDTLS_RSA_PKCS_V15`.)

> 2\. Original alg: v2.1 (OAEP or PSS) with a concrete hash  
> 3\. Original alg: v2.1 (OAEP or PSS) with a ANY_HASH

Note that OAEP doesn't allow `ANY_HASH`. The wildcard is only valid for signature algorithms.

I guess the obvious choice is to set `rsa->padding = MBEDTLS_RSA_PKCS_V21` and `rsa->hash_id` to the hash, or 0 for `ANY_HASH`, and then the pk functions do whatever they do now.

>  sign() and even sign_ext() don't require the use to specify a hash, as they both have the following in their documentation:

I don't think that's a problem. If no hash is specified at any point, you'll get an error when signing. The goal here is to reconcile two APIs where they can be reconciled, with a focus on good cases, not to design a more robust API.

----

> What are these functions supposed to do with a context where `mbedtls_rsa_set_padding(mbedtls_pk_rsa(...), MBEDTLS_RSA_PKCS_V21, ...)` was called? That's unclear to me from the documentation. Even the intent is unclear to me from memories. Also, I doubt we are testing any of this.

Do you mean an <s>African swallow or a European swallow</s> `MBEDTLS_USE_PSA_CRYPTO` build or default build? Because from [code inspection](https://github.com/Mbed-TLS/mbedtls/blob/mbedtls-3.5.0/library/pk_wrap.c#L359) I think they're different. And `grep mbedtls_rsa_set_padding tests/suites/test_suite_pk.function` confirms that we aren't testing it.

```
#if defined(MBEDTLS_USE_PSA_CRYPTO)
…
    return mbedtls_pk_psa_rsa_sign_ext(PSA_ALG_RSA_PKCS1V15_SIGN(
```
#else /* MBEDTLS_USE_PSA_CRYPTO */
…
    return mbedtls_rsa_pkcs1_sign(rsa, f_rng, p_rng,
…
#endif /* MBEDTLS_USE_PSA_CRYPTO */
```

I think this counts as a bug.

gilles-peskine-arm commented:
[Bug confirmed](https://github.com/Mbed-TLS/mbedtls/issues/8824)

mpg commented:
> What are these functions supposed to do with a context where `mbedtls_rsa_set_padding(mbedtls_pk_rsa(...), MBEDTLS_RSA_PKCS_V21, ...)` was called? That's unclear to me from the documentation. Even the intent is unclear to me from memories. Also, I doubt we are testing any of this.

Thanks for looking into this. I think your work also confirms that we were not testing any of this, otherwise the tests would have caught the discrepancy between `USE_PSA` and `!USE_PSA`. I agree this is a bug, as even if it's not explicitly documented, the behaviour has been there since day 1 and meets a reasonable expectation.

I suggest we pause this and resolve #8824 first. Otherwise it will be hard to test proper behaviour of the copied keys here.

mpg reviewed (commented):
Thanks for addressing my feedback, I'm happy with the result, except of course for the point about the behaviour that was still under discussion.

For that, I think we'll want to address #8824 first, and then we'll probably want more tests about the fact that `pk_sign`, `verify`, `decrypt`, `encrypt` behave the way we intend them to. 

Practically, I think we'll want to ensure they do the same as doing the same operation with the original PSA key - for deterministic schemes, we can do the operation with both keys and compare the results, for non-deterministic things we can do one operation with one key and the other with the other (like, sign with the PSA key and verify with the PK key and vice-versa).

(Then we probably want special cases for `ANY_HASH` which should work if a proper hash identifier is passed to the PK function and fail otherwise.)

gilles-peskine-arm commented on tests/suites/test_suite_pk.function:
```diff
@@ -1924,3 +1971,202 @@ exit:
+                                         &pk_ctx2, md_for_test, in_buf, in_buf_len,
+                                         out_buf, out_buf_len), 0);
+    } else {
+        TEST_EQUAL(mbedtls_pk_sign(&pk_ctx, md_for_test, in_buf, in_buf_len,
```
(By the way, well done for having this consistency testing, I didn't think of that when I did import_from_psa.)

gilles-peskine-arm replied:
On further reflection, if we're confident about the internal consistency of each key, then it's enough to compare the public keys, and I implemented that in https://github.com/Mbed-TLS/mbedtls/pull/8807. I suggest we each keep doing what we're doing, and we can reconcile the testing in a follow-up.

gilles-peskine-arm commented on tests/suites/test_suite_pk.function:
```diff
@@ -1924,3 +1971,202 @@ exit:
+/* END_CASE */
+
+/* BEGIN_CASE depends_on:MBEDTLS_PSA_CRYPTO_C*/
+void pk_copy_from_psa_success(data_t *priv_key_data, int key_type_arg,
```
What about copying a public key?

valeriosetti replied:
Isn't this already done when I create `pk_ctx2`?
For that one I start from the private key held on PSA side, I export its public counterpart and I create a new slot still in PSA, then I call `mbedtls_pk_copy_from_psa` on this public key.

gilles-peskine-arm replied:
Ah, right, sorry, I missed that. I've only looked superficially so far, mostly to compare the testing with `import_into_psa`, which I only lately realized is similar.

Please rename `pk_ctx2` to something like `pk_pub`, to make this clearer.

gilles-peskine-arm commented on tests/suites/test_suite_pk.function:
```diff
@@ -1924,3 +1971,202 @@ exit:
+     * counterpart generated above.  */
+    pub_key_id = pk_psa_pub_key_from_priv(priv_key_id, pub_key_type, key_usage, key_alg, key_bits);
+    TEST_EQUAL(mbedtls_pk_copy_from_psa(pub_key_id, &pk_ctx2), 0);
+
```
Please do an internal consistency check: `mbedtls_pk_check_pair`. (Only applies for a key pair, the pk module doesn't have a corresponding function for public keys.)

mpg commented:
@valeriosetti Now that #8824 is fixed (thanks!) I think this can be rebased and the tests expanded to ensure sign, verify, decrypt, encrypt use the intended padding for RSA keys.

gilles-peskine-arm commented:
If you're going to rebase this anyway, please wait until the merge of https://github.com/Mbed-TLS/mbedtls/pull/8807 is in (should be a couple of hours). I expect some trivial conflicts in the tests.

valeriosetti commented:
> If you're going to rebase this anyway, please wait until the merge of #8807 is in (should be a couple of hours). I expect some trivial conflicts in the tests.

Thanks for the tip! I'm actually already working on this: rebase is done and I'm currently working on enhancing tests. I might push the current status before doing another rebase just to check what the CI thinks.

valeriosetti commented:
https://github.com/Mbed-TLS/mbedtls/pull/8774/commits/f17ad5ecc9098cbf4ac0d1a141c36ff3beb30f69 includes a fix that I forgot/didn't see in #8826.

mpg commented:
> [f17ad5e](https://github.com/Mbed-TLS/mbedtls/commit/f17ad5ecc9098cbf4ac0d1a141c36ff3beb30f69) includes a fix that I forgot/didn't see in #8826.

Thanks for fixing that. But that means we also have a test gap, as this should have been caught by tests.

valeriosetti commented:
> > [f17ad5e](https://github.com/Mbed-TLS/mbedtls/commit/f17ad5ecc9098cbf4ac0d1a141c36ff3beb30f69) includes a fix that I forgot/didn't see in #8826.
> 
> Thanks for fixing that. But that means we also have a test gap, as this should have been caught by tests.

New tests that I added here surely do catch that (that's how I discovered it :D), but I can investigate why this didn't happen in tests added in #8826 and fix that as well.

gilles-peskine-arm reviewed (changes requested):
I've done a first pass of review, covering everything (doc, tests) except the implementation of the new function. I'll do a second pass hopefully next week.

valeriosetti commented on tests/suites/test_suite_pk.function:
```diff
@@ -2196,3 +2246,242 @@ exit:
+    mbedtls_pk_init(&pk_pub);
+    PSA_INIT();
+
+    /* Generate a private key in PSA and create a PK context from it. */
```
This probably dates back to the original implementation in which I really used to generate the keys to be used in the test. Then I abandoned this approach since it seemed to me that it was imposing too many constraint to the test (i.e. having support for key's generation). Unfortunately I forgot to update the comment after that change :(

valeriosetti commented on tests/suites/test_suite_pk.function:
```diff
@@ -2196,3 +2246,242 @@ exit:
+    /* Test #4: in case of RSA key pair try also encryption/decryption.
+     * Hash algorithm does not matter when padding mode is PKCS1 v1.5, whereas
+     * it must be not null or any in case of v2.1 (OAEP). */
+    int test_encryption = (PSA_ALG_IS_RSA_PKCS1V15_SIGN(key_alg) ||
+                           (key_alg == PSA_ALG_RSA_PKCS1V15_CRYPT) ||
+                           is_psa_hash_alg_specified);
+    if ((key_type == PSA_KEY_TYPE_RSA_KEY_PAIR) && test_encryption) {
```
If I remember correctly this extra check was needed because there were some test failing using the ANY_HASH alg. However as you pointed out in a [later comment](https://github.com/Mbed-TLS/mbedtls/pull/8774#discussion_r1499891284) that's probably due to some wrong configuration. I guess that removing that case will help also optimizing this check here.

mpg reviewed (changes requested):
Thanks for improving the tests! Still got a few more requests though :)

valeriosetti commented on library/pk.c:
```diff
@@ -27,14 +27,18 @@
 #endif
 
-#if defined(MBEDTLS_PSA_CRYPTO_CLIENT)
+#if defined(MBEDTLS_PSA_CRYPTO_C)
```
I changed this guard when I also changed the one for `copy_from_psa`. I thought that it was better to have them aligned (then resetting both of them to `CRYPTO_CLIENT` whenever possible), but no problem in reverting this one even now.

valeriosetti commented on tests/suites/test_suite_pk.function:
```diff
@@ -2375,63 +2357,118 @@ void pk_copy_from_psa_success(data_t *priv_key_data, int key_type_arg,
-    test_encryption = ((PSA_ALG_GET_HASH(key_alg) != 0) &&
-                       (PSA_ALG_GET_HASH(key_alg) != PSA_ALG_ANY_HASH));
-#endif
+#if defined(MBEDTLS_PSA_UTIL_HAVE_ECDSA)
```
Ops, I totally forgot about RSA when placing this guard it seems :/

valeriosetti commented on tests/suites/test_suite_pk.function:
```diff
@@ -2375,63 +2357,118 @@ void pk_copy_from_psa_success(data_t *priv_key_data, int key_type_arg,
                                        out_buf, sizeof(out_buf), &out_buf_len,
                                        mbedtls_test_rnd_std_rand, NULL), 0);
+    } else {
+        TEST_EQUAL(mbedtls_pk_sign(&pk_priv, md_for_test, in_buf, in_buf_len,
```
After Gilles' last review, `pk_sign_ext()` is basically only used in case of PSS + ANY_HASH, because we need to pass an MD alg for the signature. Should I remove this case as well as I did with OAEP+ANY_HASH?

valeriosetti replied:
Mmmm I think I found another disparity between `!USE_PSA` and `USE_PSA` cases. Tests with `PSS_ANY_SALT + ANY_HASH)` and `PSS + ANY_HASH` have:
- `pk_sign()` failing when `!USE_PSA`
- `pk_sign()` succesful when `USE_PSA`

This doesn't look correct to me. Wdyt?

valeriosetti replied:
Digging a bit more on this I found that the call sequence in the `!USE_PSA` case is:
`rsa_sign_wrap->mbedtls_rsa_pkcs1_sign->mbedtls_rsa_rsassa_pss_sign->rsa_rsassa_pss_sign`
and `rsa_rsassa_pss_sign` performs the check on RSA context `hash_id` to verify that it is not `MBEDTLS_MD_NONE`. Skipping this check, i.e. calling directly `rsa_rsassa_pss_sign_no_mode_check` makes `pk_sign()` pass just like the `USE_PSA` case.

gilles-peskine-arm replied:
`PSS(ANY_HASH)` is just a wildcard policy, it isn't a valid algorithm. So with the key resulting from `copy_from_psa`, the hash can only come from the `md_alg` parameter of `mbedtls_pk_{sign,verify}``, which can't be `MBEDTLS_MD_NONE`. Under the hood, that means `set_padding_mode(PKCS_V21, MD_NONE).` on the RSA context.

valeriosetti replied:
Ok, thanks for the explanation. I thought that `PSS(ANY_HASH)` could be valid with the following explanation:
- at PK build time the user sets `ANY_HASH` (which unfortunately translates to `MD_NONE` which does not sound the same) because he/she is not interested in setting any hash since the beginning;
- then he/she specifies the MD alg when doing the signature.

However I guess that `pk_sign_ext()` is the one that must be used for the above purpose. Isn't it?
So in this case it's the `USE_PSA` that is wrong, correct?

gilles-peskine-arm replied:
`MD_NONE` means “no hash specified”, which has different meanings in different cases:

* In an RSA context, it means “I'm not saying, you'll have to pass a proper hash algorithm when you call an operation that needs one”.
* For pk sign/verify for an RSA key whose padding mode is set to PKCS#1v1.5, it means “`hash` contains a DigestInfo, not a hash”.
* For pk sign/verify for an RSA key whose padding mode is set to PKCS#1v2.1 (so PSS), it means “`use the  hash algorithm in the context (leading to an error if that is also `MD_NONE`).

``PSA_ALG_ANY_HASH` in the policy means “no hash specified” when creating the key, so it corresponds to `MBEDTLS_MD_NONE` in the RSA context.

The non-ext pk sign/verify function have a `md_alg` parameter so you don't need to use the ext functions to specify a hash.

> So in this case it's the USE_PSA that is wrong, correct?

I guess? The choice of hash algorithm has to come from somewhere. (For PSS: for v1.5 `MD_NONE` is valid as discussed above.) If there's no algorithm in the key policy and none passed to `mbedtls_pk_sign`, where does the `USE_PSA_CRYPTO` code get one?

gilles-peskine-arm replied:
If `pk_sign` and friends receive a hash algorithm that isn't NONE, it always overrides the value in the context. I thought we were talking about the case of `pk_sign(..., MBEDTLS_MD_NONE, ...)`.

valeriosetti replied:
Sorry, it's my fault that I didn't explain well what the issue was about. However I've just pushed a [commit](https://github.com/Mbed-TLS/mbedtls/pull/8774/commits/b2fca8df1d2076087be6669bfc346c843533d8a0) which should make the CI happy.

valeriosetti commented on tests/suites/test_suite_pk.function:
```diff
@@ -2375,63 +2357,118 @@ void pk_copy_from_psa_success(data_t *priv_key_data, int key_type_arg,
+     *
+     * Note: mbedtls_pk_verify_ext() is tested only for RSA keys with PKCS1 v2.1
+     *      padding. For other key types this functions works the same as
+     *      mbedtls_pk_verify().
```
Differently from the sign procedure, this was only to prove that `verify_ext()` is working as documented. If you think this is unnecessary I can remove it ;)

valeriosetti commented:
@gilles-peskine-arm @mpg thanks for your reviews! I think I addressed all the comments so the PR is (almost) ready for new review.
I said "almost ready" because I don't expect the CI to fully pass due to [this problem](https://github.com/Mbed-TLS/mbedtls/pull/8774#discussion_r1500949305) that I found. I will investigate on that next week.

valeriosetti commented:
I had to rebase to solve a conflict with development in `pk_wrap.c`. The CI was basically fine before, only having a failure in `API-ABI` check, so I think the PR is ready for review.

gilles-peskine-arm commented:
The internal CI glitched in a DTLS test. We can ignore that.

mpg reviewed (commented):
Incremental review. I'll do another pass later to check consistency of the end result, but releasing those comments in the meantime.

mpg commented on include/mbedtls/pk.h:
```diff
@@ -394,35 +394,36 @@ int mbedtls_pk_setup_opaque(mbedtls_pk_context *ctx,
  *                  This key:
- *                  - must have PSA_KEY_USAGE_EXPORT attribute set and
- *                  - must be a either a RSA or EC (DH is not managed in PK) and
+ *                  - must be exportabel and
```
Fixed in a later commit.

mpg commented:
Actually I'm wondering if the testing approach is overkill or not: doing actual signatures and verifications is computationally intensive. Instead we could just:
- [already done] check the key type and, for RSA, padding mode and hash
- [could be added] export the public key and check that it matches the public key, like in #8807.

Then rely on existing tests (and perhaps extend them*) to ensure that an RSA key with a given padding and hash behaves as expected when using sign, verify, encrypt, decrypt. *For the case of `set_padding(..., MD_NONE)` where this PR caught a discrepancy that wasn't caught by existing tests.

I mean, this is supposed to be unit tests, so it would make sense to test one thing at a time: here we'd be only testing that `copy_from_psa()` sets the type and padding and hash as expected - and elsewhere test how sign, verify, encrypt, decrypt, bhave as expected with a given padding and hash.

Wdyt?

valeriosetti commented on tests/suites/test_suite_pk.data:
```diff
@@ -1460,52 +1460,48 @@ pk_copy_from_psa_fail:
 Copy from PSA: valid EC (SECP_R1_256 + SHA_256)
 depends_on:MBEDTLS_ECP_HAVE_SECP256R1:MBEDTLS_PK_CAN_ECDSA_SIGN:MBEDTLS_PK_CAN_ECDSA_VERIFY:MBEDTLS_MD_CAN_SHA256
-pk_copy_from_psa_success:"587CF7C57EB7C6254CBF80CC59846521B4FBCBA8BC4B362A9B043F0DEB49CCA1":PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1):256:PSA_KEY_USAGE_SIGN_HASH | PSA_KEY_USAGE_VERIFY_HASH | PSA_KEY_USAGE_EXPORT:PSA_ALG_ECDSA(PSA_ALG_SHA_256)
+pk_copy_from_psa_success:"587CF7C57EB7C6254CBF80CC59846521B4FBCBA8BC4B362A9B043F0DEB49CCA1":PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1):PSA_ALG_ECDSA(PSA_ALG_SHA_256)
```
Yes, I think this is possible. However this goes back to the original implementation in which I used to generate the keys, which I then abandoned in favor of an explicit key in order to reduce requirements for the test.
On the pro side of your suggested change there is surely a wider support for any type of EC key, without sticking to a particular curve. I will give this a try.

valeriosetti replied:
Mmm wait, I just realized that the `ONE_FAMILY` macro can also pick Montgomery keys, but as far as I know those are not allowed to do ECDSA and therefore they might require further tricks for the `pk_copy_from_psa_success` test function :/

mpg replied:
Oh yes, sorry, missed that. Then I think I'm more inclined to keep the current version, because generating keys comes with a computational cost - and as you say, also more dependencies (`GENPRIME` for RSA).

gilles-peskine-arm replied:
> generating keys comes with a computational cost

I'd like to replace it by a hard-coded RSA key, or an RSA key that's cached between instances. But that's not going to happen before 3.6.

gilles-peskine-arm replied:
Filed as https://github.com/Mbed-TLS/mbedtls/issues/8871 — let's handle that after the 3.6 code freeze.

gilles-peskine-arm commented:
> Then rely on existing tests (and perhaps extend them*) to ensure that an RSA key with a given padding and hash behaves as expected when using sign, verify, encrypt, decrypt.

Yes, that's the “exercise key” thing with PSA keys. But only testing metadata misses a bug you found in #8807 where the converted key was valid, but not the same key as the original. Hence Valerio's **cross-check** like sign-with-original/verify-with-copy, or the cheaper comparison of public keys in my PR.

mpg commented:
>  Hence Valerio's cross-check like sign-with-original/verify-with-copy, or the cheaper comparison of public keys in my PR.

Agreed, we should have one of them, I don't think I was suggesting otherwise. My question was which one: do we keep the computationally expansive cross-check, or move to the cheaper comparison of public keys?

mpg commented:
Note: I'm now inclined to keep the existing checks, because they're safe and existing. Making the test suite less computationally intensive for equivalent coverage would be an optimization that can be done later (and there might be  bigger wins than this, for example avoiding key generation).

valeriosetti commented on tests/suites/test_suite_pk.function:
```diff
@@ -2400,52 +2399,40 @@ void pk_copy_from_psa_success(data_t *priv_key_data, int key_type_arg,
-    /* Test #3: check sign/verify interoperability also in the opposite direction:
+    /* Test #4: check sign/verify interoperability also in the opposite direction:
      * sign with PSA and verify with PK. Key's policy must include a valid hash
      * algorithm (not any).
```
> I'm sure we have a macro or function somewhere that just picks a hash that's enabled, but I don't remember where right now

Sorry but I didn't find that so I re-implemented a [local version](https://github.com/Mbed-TLS/mbedtls/pull/8774/commits/f7eca7a73f22598ebbd80cf337118827ea05b10a).

gilles-peskine-arm replied:
For a PSA supported there's `KNOWN_SUPPORTED_HASH_ALG` in `test/psa_exercise_key.h`. If you need a hash that's also supported in the legacy API I don't think we had one yet.

gilles-peskine-arm commented:
I've filed follow-up issues to improve the test code:

* https://github.com/Mbed-TLS/mbedtls/issues/8836
* https://github.com/Mbed-TLS/mbedtls/pull/8774

mpg reviewed (dismissed):
LGTM, thanks!

Left one suggestion for improvement, but it's not worth doing another cycle just for that.

mpg reviewed (dismissed):
LGTM ;)

gilles-peskine-arm reviewed (changes requested):
Mostly looks good at 8b7c95ef90f5416e61015b319ab195cbd8d5d71b, with a couple of things to address (plus a rebase) and a few more to preferably address.

gilles-peskine-arm commented on library/pk.c:
```diff
@@ -1357,6 +1361,116 @@ mbedtls_pk_type_t mbedtls_pk_get_type(const mbedtls_pk_context *ctx)
+        return MBEDTLS_ERR_PK_BAD_INPUT_DATA;
+    }
+
+    status = psa_export_key(key_id, exp_key, sizeof(exp_key), &exp_key_len);
```
I've gone and implemented this in https://github.com/Mbed-TLS/mbedtls/pull/8899. I made it a separate function because it was quick to do and it matches how the pk parse/write functions are designed.

valeriosetti commented on tests/suites/test_suite_pk.function:
```diff
@@ -425,7 +440,110 @@ exit:
+    type = psa_get_key_type(&attributes);
+    usage = psa_get_key_usage_flags(&attributes);
+    alg = psa_get_key_algorithm(&attributes);
+    psa_reset_key_attributes(&attributes);
```
If I'm not wrong, resetting is also useful to clear the key ID from the attributes, otherwise the following import would fail. But I can give this a try

valeriosetti commented:
I think I addressed all the comments, but I just run some quick test locally. I'm not sure the CI will be happy on the first run, so I suggest to wait a bit before starting a new review cycle.

valeriosetti commented:
> Lots of compilation errors, I think the `full` config is broken.

Yes, a rebase error indeed. Now the OpenCI fully passed so the PR is ready for review :)

gilles-peskine-arm commented on library/pk.c:
```diff
@@ -1374,4 +1378,113 @@ mbedtls_pk_type_t mbedtls_pk_get_type(const mbedtls_pk_context *ctx)
+
+        if (PSA_ALG_IS_RSA_OAEP(alg_type) || PSA_ALG_IS_RSA_PSS(alg_type)) {
+            ret = mbedtls_rsa_set_padding(mbedtls_pk_rsa(*pk), MBEDTLS_RSA_PKCS_V21, md_type);
+        } else {
+            ret = mbedtls_rsa_set_padding(mbedtls_pk_rsa(*pk), MBEDTLS_RSA_PKCS_V15, md_type);
```
If support for PKCS#1v1.5 is disabled, this call fails. But we're doing it even if `alg_type == 0`. It's an edge case so not a blocker. It does show we're missing some test coverage.

I discovered this while working on https://github.com/Mbed-TLS/mbedtls/pull/8899 (a test case for the new function happened to cover it). Since I'm already on it, I'll fix the implementation and add tests there.

valeriosetti commented on tests/suites/test_suite_pk.data:
```diff
@@ -1451,3 +1455,111 @@ pk_import_into_psa_opaque:PSA_KEY_TYPE_ECC_KEY_PAIR(MBEDTLS_TEST_PSA_ECC_ONE_FAM
+Copy from PSA: use wrong parameters
+pk_copy_from_psa_fail:
+
+Copy from PSA: accelerated key only, not available as built-in
```
I just realized that the logic for EC keys is slightly different than the RSA one. In RSA we can have driver support, but not `RSA_C` and therefore test using `pk_copy_from_psa_builtin_fail` works.
For EC keys instead I should disable `MBEDTLS_PK_HAVE_ECC_KEYS` to get the same effect (`mbedtls_pk_copy_from_psa` returning `MBEDTLS_ERR_PK_BAD_INPUT_DATA`) which means disabling EC keys both as builtin and driver support.

valeriosetti commented on tests/suites/test_suite_pk.data:
```diff
@@ -1459,8 +1459,13 @@ pk_import_into_psa_opaque:PSA_KEY_TYPE_ECC_KEY_PAIR(MBEDTLS_TEST_PSA_ECC_ONE_FAM
+pk_copy_from_psa_builtin_fail:PSA_KEY_TYPE_RSA_KEY_PAIR:PSA_ALG_RSA_PKCS1V15_SIGN(PSA_ALG_SHA_256)
+
+Copy from PSA: EC SECP256R1 accelerated only, not available as built-in
+depends_on:MBEDTLS_PSA_ACCEL_ECC_SECP_R1_256:MBEDTLS_PSA_ACCEL_KEY_TYPE_ECC_KEY_PAIR_BASIC:MBEDTLS_PSA_ACCEL_ALG_ECDSA:!MBEDTLS_ECDSA_C:!MBEDTLS_ECP_DP_SECP256R1:MBEDTLS_PSA_ACCEL_ALG_SHA_256
```
The CI didn't failed but I suspect that the reason is that this test was never executed, so no disparity was found.
As I commented [above](https://github.com/Mbed-TLS/mbedtls/pull/8774#discussion_r1519896004) I think that trying to use `pk_copy_from_psa_builtin_fail` for EC keys cannot work. In a scenario in which a certain EC key is only available through a driver and not as builtin `mbedtls_pk_copy_from_psa()` uses functions that will all succeed and therefore it will succeed [^1]
Am I missing something?

[^1]: I tried using `component_test_psa_crypto_config_accel_ecc_no_ecp_at_all` and modifying `pk_copy_from_psa_builtin_fail` and its data accordingly) and indeed it succeeds.

gilles-peskine-arm replied:
> In a scenario in which a certain EC key is only available through a driver and not as builtin **mbedtls_pk_copy_from_psa() uses functions that will all succeed** and therefore it will succeed

Oh, I see. I'd missed that part. So indeed we don't need a separate case for ECC, because it's handled by `pk_copy_from_psa_success` test cases.

I prefer the RSA test case the way it is now, can you please keep it and replace the ECC test case by a comment explaining that there's no ECC test case because of what you just wrote?

valeriosetti replied:
Sure, I can revert https://github.com/Mbed-TLS/mbedtls/pull/8774/commits/7dc2bf2cb0b320e4a43c1978951b1fba02412f0e and add the comment ;)

mpg commented on tests/suites/test_suite_pk.function:
```diff
@@ -434,19 +552,12 @@ exit:
-    const psa_key_type_t type =
-        PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1);
-    const size_t bits = 256;
+    mbedtls_svc_key_id_t key;
```
This removed the initialization `= MBEDTLS_SVC_KEY_ID_INIT`. I don't think that was intentional, probably an accident in resolving conflicts. Please add the initializer back. (Multiple occurrences in this file.)

mpg reviewed (changes requested):
Looks good to me except for undoing an improvement from a previous PR while resolving conflicts.

mpg reviewed (approved):
Looks all good to me now, thanks!

gilles-peskine-arm reviewed (approved):
LGTM
